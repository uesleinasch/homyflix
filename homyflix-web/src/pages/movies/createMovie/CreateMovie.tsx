import React, { useCallback } from 'react';
import { useForm } from 'react-hook-form';
import { zodResolver } from '@hookform/resolvers/zod';
import { z } from 'zod';
import { useNavigate } from 'react-router-dom';
import { useMovieOperations } from '../../../hooks/useMovieOperations';

// Schema de validação baseado nas regras da API
const createMovieSchema = z.object({
  title: z.string()
    .min(1, 'Título é obrigatório')
    .max(255, 'Título deve ter no máximo 255 caracteres'),
  release_year: z.number()
    .int('Ano deve ser um número inteiro')
    .min(1888, 'Ano deve ser maior ou igual a 1888')
    .max(new Date().getFullYear() + 5, 'Ano não pode ser muito distante no futuro'),
  genre: z.string()
    .min(1, 'Gênero é obrigatório')
    .max(100, 'Gênero deve ter no máximo 100 caracteres'),
  synopsis: z.string()
    .min(1, 'Sinopse é obrigatória'),
  poster_url: z.string()
    .url('URL inválida')
    .optional()
    .or(z.literal(''))
});

type CreateMovieFormData = z.infer<typeof createMovieSchema>;

const CreateMovie: React.FC = () => {
  const navigate = useNavigate();
  const { createMovie, loading } = useMovieOperations();

  const {
    register,
    handleSubmit,
    formState: { errors },
    setError
  } = useForm<CreateMovieFormData>({
    resolver: zodResolver(createMovieSchema),
    defaultValues: {
      title: '',
      release_year: new Date().getFullYear(),
      genre: '',
      synopsis: '',
      poster_url: ''
    }
  });

  const onSubmit = useCallback(async (data: CreateMovieFormData) => {
    try {
      // Transformar string vazia em undefined para poster_url
      const movieData = {
        title: data.title,
        release_year: data.release_year,
        genre: data.genre,
        synopsis: data.synopsis,
        poster_url: data.poster_url || null
      };

      const result = await createMovie(movieData);
      
      if (result.success && result.data) {
        navigate(`/movies/${result.data.id}`);
      } else if (result.error) {
        // Mostrar erro genérico
        setError('root', {
          type: 'manual',
          message: result.error
        });
      }
    } catch {
      setError('root', {
        type: 'manual',
        message: 'Erro ao criar filme. Tente novamente.'
      });
    }
  }, [createMovie, navigate, setError]);

  const handleCancel = useCallback(() => {
    navigate('/movies');
  }, [navigate]);

  return (
    <div>
      <h1>Criar Novo Filme</h1>

      <form onSubmit={handleSubmit(onSubmit)}>
        {/* Erro geral */}
        {errors.root && (
          <div>
            <p>{errors.root.message}</p>
          </div>
        )}

        {/* Título */}
        <div>
          <label htmlFor="title">Título *</label>
          <input
            id="title"
            type="text"
            {...register('title')}
            disabled={loading}
          />
          {errors.title && <span>{errors.title.message}</span>}
        </div>

        {/* Ano de Lançamento */}
        <div>
          <label htmlFor="release_year">Ano de Lançamento *</label>
          <input
            id="release_year"
            type="number"
            {...register('release_year', { valueAsNumber: true })}
            disabled={loading}
          />
          {errors.release_year && <span>{errors.release_year.message}</span>}
        </div>

        {/* Gênero */}
        <div>
          <label htmlFor="genre">Gênero *</label>
          <input
            id="genre"
            type="text"
            {...register('genre')}
            disabled={loading}
          />
          {errors.genre && <span>{errors.genre.message}</span>}
        </div>

        {/* Sinopse */}
        <div>
          <label htmlFor="synopsis">Sinopse *</label>
          <textarea
            id="synopsis"
            {...register('synopsis')}
            disabled={loading}
            rows={5}
          />
          {errors.synopsis && <span>{errors.synopsis.message}</span>}
        </div>

        {/* URL do Poster */}
        <div>
          <label htmlFor="poster_url">URL do Poster</label>
          <input
            id="poster_url"
            type="url"
            {...register('poster_url')}
            disabled={loading}
            placeholder="https://exemplo.com/poster.jpg"
          />
          {errors.poster_url && <span>{errors.poster_url.message}</span>}
        </div>

        {/* Botões de ação */}
        <div>
          <button
            type="submit"
            disabled={loading}
          >
            {loading ? 'Criando...' : 'Criar Filme'}
          </button>
          
          <button
            type="button"
            onClick={handleCancel}
            disabled={loading}
          >
            Cancelar
          </button>
        </div>
      </form>
    </div>
  );
};

export default CreateMovie;