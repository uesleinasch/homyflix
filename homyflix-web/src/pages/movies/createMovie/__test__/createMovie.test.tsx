import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";
import { MantineProvider } from "@mantine/core";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import { configureStore } from "@reduxjs/toolkit";
import CreateMovie from "../CreateMovie";
import movieReducer from "../../../../store/slices/movieSlice";
import authReducer from "../../../../store/slices/authSlice";

// Mock do import.meta.env (específico para este teste)
if (!(globalThis as any).import?.meta) {
  Object.defineProperty(globalThis, 'import', {
    value: {
      meta: {
        env: {
          VITE_API_URL: 'http://localhost:8000/api',
          MODE: 'test',
          DEV: true,
          PROD: false
        }
      }
    },
    configurable: true
  });
}

// Mock da API axios
jest.mock("../../../../core/auth/api", () => ({
  default: {
    defaults: { baseURL: 'http://localhost:8000/api' },
    interceptors: {
      request: { use: jest.fn() },
      response: { use: jest.fn() }
    },
    get: jest.fn(),
    post: jest.fn(),
    put: jest.fn(),
    delete: jest.fn()
  }
}));

// Mock do tokenManager
jest.mock("../../../../core/auth/tokenManager", () => ({
  tokenManager: {
    getToken: jest.fn(() => 'fake-token'),
    getCurrentUser: jest.fn(() => null),
    isTokenValid: jest.fn(() => true),
    clearToken: jest.fn()
  }
}));

// Mock dos serviços
jest.mock("../../../../core/services/movieService", () => ({
  movieService: {
    getAllMovies: jest.fn(),
    getMovieById: jest.fn(),
    createMovie: jest.fn(),
    updateMovie: jest.fn(),
    deleteMovie: jest.fn()
  }
}));

jest.mock("../../../../shared/services/authApplicationService", () => ({
  authApplicationService: {
    authenticateUser: jest.fn(),
    registerUser: jest.fn(),
    logoutUser: jest.fn(),
    refreshUserToken: jest.fn()
  }
}));

// Mock do react-router-dom
const mockNavigate = jest.fn();
const mockUseParams = jest.fn(() => ({}));

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockNavigate,
  useParams: () => mockUseParams(),
}));

// Mock do hook customizado - MAIS IMPORTANTE PARA EVITAR O LOOP
const mockCreateNewMovie = jest.fn();
const mockUpdateExistingMovie = jest.fn();
const mockLoadMovieById = jest.fn();

jest.mock("../../../../core/hooks/useMovieOperations", () => ({
  useMovieOperations: jest.fn(() => ({
    createNewMovie: mockCreateNewMovie,
    updateExistingMovie: mockUpdateExistingMovie,
    loadMovieById: mockLoadMovieById,
    loading: false,
    movies: [],
    error: null,
    clearMovieError: jest.fn(),
  })),
}));

// Mock dos componentes que podem causar problemas
jest.mock("../../../../shared/components/ui/header/Header", () => {
  return function MockHeader() {
    return <div data-testid="mock-header">Header</div>;
  };
});

jest.mock("../../../../shared/components/ui/loaderScreen/LoadScreen", () => {
  return function MockLoadScreen({ isLoading, loadingText }: { isLoading: boolean; loadingText: string }) {
    return isLoading ? <div data-testid="mock-load-screen">{loadingText}</div> : null;
  };
});

jest.mock("../../../../shared/components/ui/mantineContainer/MantineContainer", () => {
  return function MockMantineContainer({ children }: { children: React.ReactNode }) {
    return <div data-testid="mock-mantine-container">{children}</div>;
  };
});

describe("CreateMovie", () => {
  beforeEach(() => {
    jest.clearAllMocks();
    mockUseParams.mockReturnValue({});
  });

  const renderWithProviders = (initialState = {}, routerProps = {}) => {
    const store = configureStore({
      reducer: {
        movies: movieReducer,
        auth: authReducer,
      },
      preloadedState: {
        movies: {
          movies: [],
          isLoading: false,
          error: null,
        },
        auth: {
          user: null,
          token: "fake-token",
          isAuthenticated: true,
          isLoading: false,
          error: null,
        },
        ...initialState,
      },
    });

    return render(
      <Provider store={store}>
        <MemoryRouter initialEntries={["/movies/create"]} {...routerProps}>
          <MantineProvider>
            <CreateMovie />
          </MantineProvider>
        </MemoryRouter>
      </Provider>
    );
  };

  it("should render create movie form without crashing", () => {
    const { container } = renderWithProviders();
    
    expect(container).toBeDefined();
    
    expect(screen.getByTestId("mock-mantine-container")).toBeInTheDocument();
  });

  it("should render in create mode when no id param is provided", () => {
    renderWithProviders();
    
    expect(mockLoadMovieById).not.toHaveBeenCalled();
  });

  it("should render in edit mode when id param is provided", () => {
    mockUseParams.mockReturnValue({ id: "123" });
    
    renderWithProviders();
    
    expect(mockLoadMovieById).toHaveBeenCalledWith(123);
  });

  it("should handle invalid id param gracefully", () => {
    mockUseParams.mockReturnValue({ id: "invalid" });
    
    const { container } = renderWithProviders();
    
    expect(container).toBeDefined();
    
    expect(mockLoadMovieById).not.toHaveBeenCalled();
  });

 
});
