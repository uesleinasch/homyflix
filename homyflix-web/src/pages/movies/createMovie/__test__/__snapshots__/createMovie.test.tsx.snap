// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateMovie should match snapshot 1`] = `
<div>
  <style
    data-mantine-styles="classes"
  >
    @media (max-width: 35.99375em) {.mantine-visible-from-xs {display: none !important;}}@media (min-width: 36em) {.mantine-hidden-from-xs {display: none !important;}}@media (max-width: 47.99375em) {.mantine-visible-from-sm {display: none !important;}}@media (min-width: 48em) {.mantine-hidden-from-sm {display: none !important;}}@media (max-width: 61.99375em) {.mantine-visible-from-md {display: none !important;}}@media (min-width: 62em) {.mantine-hidden-from-md {display: none !important;}}@media (max-width: 74.99375em) {.mantine-visible-from-lg {display: none !important;}}@media (min-width: 75em) {.mantine-hidden-from-lg {display: none !important;}}@media (max-width: 87.99375em) {.mantine-visible-from-xl {display: none !important;}}@media (min-width: 88em) {.mantine-hidden-from-xl {display: none !important;}}
  </style>
  <div
    data-testid="mock-mantine-container"
  >
    <div
      data-testid="mock-header"
    >
      Header
    </div>
    <div>
      <form>
        <style
          data-mantine-styles="inline"
        >
          .__m__-«r0»{padding-top:var(--mantine-spacing-md);padding-bottom:var(--mantine-spacing-md);padding-left:0rem;padding-right:0rem;}@media(min-width: 48em){.__m__-«r0»{padding-top:calc(6.25rem * var(--mantine-scale));padding-bottom:calc(5rem * var(--mantine-scale));padding-left:calc(18.75rem * var(--mantine-scale));padding-right:calc(18.75rem * var(--mantine-scale));}}
        </style>
        <div
          class="__m__-«r0»"
        >
          <div
            class="m_1b7284a3 mantine-Paper-root"
            style="--paper-shadow: var(--mantine-shadow-xs);"
          >
            <style
              data-mantine-styles="inline"
            >
              .__m__-«r2»{--grid-gutter:var(--mantine-spacing-md);}
            </style>
            <div
              class="m_410352e9 mantine-Grid-root __m__-«r2»"
              style="--grid-justify: center;"
            >
              <div
                class="m_dee7bd2f mantine-Grid-inner"
              >
                <style
                  data-mantine-styles="inline"
                >
                  .__m__-«r4»{--col-flex-grow:auto;--col-flex-basis:100%;--col-max-width:100%;}@media(min-width: 62em){.__m__-«r4»{--col-flex-grow:auto;--col-flex-basis:50%;--col-max-width:50%;}}@media(min-width: 75em){.__m__-«r4»{--col-flex-grow:auto;--col-flex-basis:100%;--col-max-width:100%;}}
                </style>
                <div
                  class="m_96bdd299 mantine-Grid-col __m__-«r4»"
                >
                  <div
                    class="m_46b77525 mantine-InputWrapper-root mantine-TextInput-root"
                  >
                    <label
                      class="m_8fdc1311 mantine-InputWrapper-label mantine-TextInput-label"
                      for="mantine-is6vfgval"
                      id="mantine-is6vfgval-label"
                    >
                      Título *
                    </label>
                    <div
                      class="m_6c018570 mantine-Input-wrapper mantine-TextInput-wrapper"
                      data-variant="default"
                    >
                      <input
                        aria-invalid="false"
                        class="m_8fb7ebe7 mantine-Input-input mantine-TextInput-input"
                        data-variant="default"
                        id="mantine-is6vfgval"
                        name="title"
                      />
                    </div>
                  </div>
                  <div
                    class="m_46b77525 mantine-InputWrapper-root mantine-YearPickerInput-root"
                  >
                    <label
                      class="m_8fdc1311 mantine-InputWrapper-label mantine-YearPickerInput-label"
                      for="mantine-cl3r04ha0"
                      id="mantine-cl3r04ha0-label"
                    >
                      Ano de Lançamento *
                    </label>
                    <div
                      class="m_6c018570 mantine-Input-wrapper mantine-YearPickerInput-wrapper"
                      data-multiline="true"
                      data-variant="default"
                    >
                      <button
                        aria-controls="mantine-obskp9uge-dropdown"
                        aria-expanded="false"
                        aria-haspopup="dialog"
                        aria-invalid="false"
                        class="m_6fa5e2aa m_8fb7ebe7 mantine-Input-input mantine-YearPickerInput-input"
                        data-dates-input="true"
                        data-variant="default"
                        id="mantine-cl3r04ha0"
                        type="button"
                      >
                        2025
                      </button>
                    </div>
                  </div>
                  <input
                    type="hidden"
                    value="2025-01-01"
                  />
                  <div
                    class="m_46b77525 mantine-InputWrapper-root mantine-Select-root"
                  >
                    <label
                      class="m_8fdc1311 mantine-InputWrapper-label mantine-Select-label"
                      for="mantine-wzfgu5uui"
                      id="mantine-wzfgu5uui-label"
                    >
                      Gênero *
                    </label>
                    <div
                      class="m_6c018570 mantine-Input-wrapper mantine-Select-wrapper"
                      data-pointer="true"
                      data-variant="default"
                      data-with-right-section="true"
                      style="--input-right-section-pointer-events: none;"
                    >
                      <input
                        aria-haspopup="listbox"
                        aria-invalid="false"
                        autocomplete="off"
                        class="m_8fb7ebe7 mantine-Input-input mantine-Select-input"
                        data-variant="default"
                        id="mantine-wzfgu5uui"
                        placeholder="Selecione um gênero"
                        readonly=""
                        value=""
                      />
                      <div
                        class="m_82577fc2 mantine-Input-section mantine-Select-section"
                        data-position="right"
                      >
                        <svg
                          class="m_2943220b mantine-ComboboxChevron-chevron"
                          data-combobox-chevron="true"
                          fill="none"
                          viewBox="0 0 15 15"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            clip-rule="evenodd"
                            d="M4.93179 5.43179C4.75605 5.60753 4.75605 5.89245 4.93179 6.06819C5.10753 6.24392 5.39245 6.24392 5.56819 6.06819L7.49999 4.13638L9.43179 6.06819C9.60753 6.24392 9.89245 6.24392 10.0682 6.06819C10.2439 5.89245 10.2439 5.60753 10.0682 5.43179L7.81819 3.18179C7.73379 3.0974 7.61933 3.04999 7.49999 3.04999C7.38064 3.04999 7.26618 3.0974 7.18179 3.18179L4.93179 5.43179ZM10.0682 9.56819C10.2439 9.39245 10.2439 9.10753 10.0682 8.93179C9.89245 8.75606 9.60753 8.75606 9.43179 8.93179L7.49999 10.8636L5.56819 8.93179C5.39245 8.75606 5.10753 8.75606 4.93179 8.93179C4.75605 9.10753 4.75605 9.39245 4.93179 9.56819L7.18179 11.8182C7.35753 11.9939 7.64245 11.9939 7.81819 11.8182L10.0682 9.56819Z"
                            fill="currentColor"
                            fill-rule="evenodd"
                          />
                        </svg>
                      </div>
                    </div>
                  </div>
                  <input
                    type="hidden"
                    value=""
                  />
                  <div
                    class="m_46b77525 mantine-InputWrapper-root mantine-Textarea-root"
                  >
                    <label
                      class="m_8fdc1311 mantine-InputWrapper-label mantine-Textarea-label"
                      for="mantine-zcfh9wbs9"
                      id="mantine-zcfh9wbs9-label"
                    >
                      Sinopse *
                    </label>
                    <div
                      class="m_6c018570 mantine-Input-wrapper mantine-Textarea-wrapper"
                      data-multiline="true"
                      data-variant="default"
                    >
                      <textarea
                        aria-invalid="false"
                        class="m_8fb7ebe7 mantine-Input-input mantine-Textarea-input"
                        data-variant="default"
                        id="mantine-zcfh9wbs9"
                        name="synopsis"
                        rows="5"
                      />
                    </div>
                  </div>
                  <div
                    class="m_46b77525 mantine-InputWrapper-root mantine-TextInput-root"
                  >
                    <label
                      class="m_8fdc1311 mantine-InputWrapper-label mantine-TextInput-label"
                      for="mantine-zabm1t8i5"
                      id="mantine-zabm1t8i5-label"
                    >
                      URL do Poster
                    </label>
                    <div
                      class="m_6c018570 mantine-Input-wrapper mantine-TextInput-wrapper"
                      data-variant="default"
                    >
                      <input
                        aria-invalid="false"
                        class="m_8fb7ebe7 mantine-Input-input mantine-TextInput-input"
                        data-variant="default"
                        id="mantine-zabm1t8i5"
                        name="poster_url"
                        placeholder="https://exemplo.com/poster.jpg"
                        type="url"
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </form>
    </div>
  </div>
</div>
`;
